WIP: Add a Tutorial and Wiki section to the documentation (#79)
davenverse/shellfish: docs/examples/index.md

# Working with Files

Here you'll find a curated collection of code examples that show the library in action.

> ## Scores

Reviewer: Can you add for each examples a few words on what you're doing in the example? At the moment this is just a bunch of code examples w/o explanations.

Author: Working on it!

# Working with Files

Here you will find a curated collection of code examples that show the library in action.

## Scores

Imagine we have a file with scores in the following format (`<name>:<score>`):

```text
michel:21839
zeta:9929
thanh:20933
tonio:12340
arman:6969
gabriel:32123
```

We also want to read the file, parse the scores, and print them to the console. We also want to append a new score to the file.

For that, you will first need to create a representation of the scores in Scala, that is via a `case class`:

```scala 3
case class Score(name: String, score: Int):
  def show: String = s"$name:$score"
```
The `show` function is just a `toString()` method, but in the world of the Cats library!

> Then you will need to create a function that parses a string into a `Score`:

Reviewer: ```suggestion
and then a function that parses a string into a `Either[Throwable, Score]`:
```

```scala 3
def parseScore(strScore: String): Either[Throwable, Score] =
  Either.catchNonFatal(                                               // (1)
    strScore.split(':') match
      case Array(name, score) => Score(name, score.toInt)   
      case _                  => Score("Cant parse this score", -1)   // (2)
  )
```

> The method is going to return an `Either` type, which is a type that can be either a `Left` or a `Right`. In this case, we are using it to represent a success or a failure in parsing the score. It is very similar to the `Result` type in Rust! Here the function is doing two things:

Reviewer: ```suggestion
The method is going to return a `Left` or a `Right` according to to the outcome of the score parsing. The function consists in two steps:
```

1. `Either.catchNonFatal` is a function that catches exceptions and wraps them in a `Left` value.
2. If the score cannot be parsed, we return a default score.

> After that, we start creating our script: 

Reviewer: ```suggestion
We can now create our script: 
```

# Working with Files

Here you will find a curated collection of code examples that show the library in action.

## Scores

Imagine we have a file with scores in the following format (`<name>:<score>`):

```text
michel:21839
zeta:9929
thanh:20933
tonio:12340
arman:6969
gabriel:32123
```

> We also want to read the file, parse the scores, and print them to the console. We also want to append a new score to the file.

Reviewer: ```suggestion
We want to read the file, parse the scores, print them to the console and also append new scores to the file.
```

> For that, you will first need to create a representation of the scores in Scala, that is via a `case class`:

Reviewer: ```suggestion
In order to perform these operations we first need to create a representation of the score in Scala:
```

```scala 3
case class Score(name: String, score: Int):
  def show: String = s"$name:$score"
```

> The `show` function is just a `toString()` method, but in the world of the Cats library!

Reviewer: ```suggestion
```

end Scores
```

@:choice(static)

```scala 3 mdoc:reset
import cats.syntax.all.*
import cats.effect.{IO, IOApp}

import fs2.io.file.Path

> import shellfish.os.FilesOs.*

Reviewer: I would avoid using `FileOs` as a wildcard import, as functions will be added in scope and they might seem "magically" present in scope (IMHO). I'll like to make people repeat the usage of the `FileOs` identifier for each function call, so that they will be able to properly distinguish which are std lib functions and which belong to the lib.

Example.
```diff
-  lines  <- readLines(path)
+  lines  <- FileOs.readLines(path)
```

Reviewer: This suggestion holds FORALL the code snippet you wrote.
